### PR TITLE
fix: remove deleted resources from local store

### PR DESCRIPTION
**Pull request by zhaohuabing**
Adsc doesn't remove deleted resources when receving xDS pushing, resulting outdated resources in the local store

Signed-off-by: huabing zhao <huabingzhao@tencent.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Should we add some tests for this?

File: pkg/adsc/adsc.go [hunk 1]
```diff
@@ -474,32 +474,38 @@ func (a *ADSC) handleRecv() {
 		clusters := []*cluster.Cluster{}
 		routes := []*route.RouteConfiguration{}
 		eds := []*endpoint.ClusterLoadAssignment{}
-		for _, rsc := range msg.Resources { // Any
-			a.VersionInfo[rsc.TypeUrl] = msg.VersionInfo
-			valBytes := rsc.Value
-			switch rsc.TypeUrl {
-			case v3.ListenerType:
+		a.VersionInfo[msg.TypeUrl] = msg.VersionInfo
+		switch msg.TypeUrl {
```
Comment (hzxuzhonghu) [inline pkg/adsc/adsc.go:478]: We can move https://github.com/istio/istio/pull/29296/files#diff-b5f677f06a05070210449a3f61f748b01824b494d5cf4116f88e50606c621d47R515-R523  here

File: pkg/adsc/adsc.go [hunk 2]
```diff
@@ -1096,51 +1102,76 @@ func (a *ADSC) GetEndpoints() map[string]*endpoint.ClusterLoadAssignment {
...
+			continue
 		}
-	}
-	if a.LocalCacheDir != "" {
-		strResponse, err := json.MarshalIndent(val, "  ", "  ")
-		if err != nil {
-			return err
+		received[val.Namespace+"/"+val.Name] = val
+
+		val.GroupVersionKind = config.GroupVersionKind{Group: gvk[0], Version: gvk[1], Kind: gvk[2]}
```
Comment (hzxuzhonghu) [inline pkg/adsc/adsc.go:1139]: duplicate with L1114

File: pkg/adsc/adsc.go [hunk 3]
```diff
@@ -1096,51 +1102,76 @@ func (a *ADSC) GetEndpoints() map[string]*endpoint.ClusterLoadAssignment {
...
+		cfg := a.Store.Get(val.GroupVersionKind, val.Name, val.Namespace)
+		if cfg == nil {
+			_, err = a.Store.Create(*val)
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
```
Comment (hzxuzhonghu) [inline pkg/adsc/adsc.go:1144]: Expect the logs to be different

File: pkg/adsc/adsc.go [hunk 4]
```diff
@@ -1096,51 +1102,76 @@ func (a *ADSC) GetEndpoints() map[string]*endpoint.ClusterLoadAssignment {
...
+				val.GroupVersionKind.Kind+"."+val.Namespace+"."+val.Name+".json", strResponse, 0644)
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
+			}
 		}
 	}
 
-	return nil
+	// remove deleted resources from cache
+	for _, config := range existingConfigs {
```
Comment (hzxuzhonghu) [inline pkg/adsc/adsc.go:1169]: This is not safe, if an existing config is updated/unchanged, this will delete it

Comment (zhaohuabing) [reply pkg/adsc/adsc.go:1169]: Name and namespace can't be updated, so updated configs won't be deleted.
xDS is using full push, so unchanged configs won't be deleted.

Comment (hzxuzhonghu) [reply pkg/adsc/adsc.go:1169]: The existingConfigs may contain unchanged/updated configs, so should move this logic before calling Create/Update

**Comment by zhaohuabing** [comment]
> Should we add some tests for this?

Tests have been added for the change.

File: pkg/adsc/adsc.go (hunk 2 above)
Comment (zhaohuabing) [reply pkg/adsc/adsc.go]: fixed

File: pkg/adsc/adsc.go (hunk 4 above)
Comment (zhaohuabing) [reply pkg/adsc/adsc.go:1169]: I have added a test case to cover this scenario:  https://github.com/istio/istio/pull/29296/files#diff-1875cf3befe58910cd1c13e4a7e224e0d5e136f9cf6ff1bdea1d428a38b7eb8fR350

File: pkg/adsc/adsc.go (hunk 3 above)
Comment (zhaohuabing) [reply pkg/adsc/adsc.go:1144]: fixed

File: pkg/adsc/adsc.go (hunk 1 above)
Comment (zhaohuabing) [reply pkg/adsc/adsc.go:478]: I didn't get it. Could you please point me to the codes?

Comment (hzxuzhonghu) [reply pkg/adsc/adsc.go:478]: I mean the two switch case can be merged

File: pkg/adsc/adsc.go (hunk 4 above)
Comment (hzxuzhonghu) [reply pkg/adsc/adsc.go:1169]: I missed that you have excluded the resources received

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**